GitLab Integration

aaronkirkham opened the pull request:
Support for GitLab API (v4) as requested in #110.

This will also support self-hosted GitLab instances, the URL must still contain the words "gitlab" (tested using a subdomain https&#58;//gitlab.mydomain.tld).

DavidAnderson684 commented:
This'll be a great addition, as Gitlab seems to be getting more and more popular.

> the URL must still contain the words "gitlab" 

That makes it unusable for me, as I didn't do that with mine. Why not add a method to allow Gitlab to be indicated explicitly? Better than unreliable guessing.

YahnisElsts commented:
I also dislike the self-hosted repository detection magic. It seems too unreliable.

If you have a self-hosted repository, I would recommend to create the PUC instances directly instead of using the factory. Something like this:

```php
$updateChecker = new Puc_v4p2_Vcs_PluginUpdateChecker(
	new Puc_v4p2_Vcs_GitHubApi('http://gitlab.example.com/'),
	__FILE__,
	'example-slug'
);
```

YahnisElsts commented on Puc/v4p2/Vcs/GitLabApi.php:
```diff
@@ -0,0 +1,285 @@
+		 */
+		public function getLatestCommit( $filename, $ref = 'master' ) {
+			throw new LogicException( 'The ' . __METHOD__ . ' method is not implemented and should not be used.' );
+		}
```
I think you can safely remove the getLatestCommit method. It's not actually used by any PUC code since version 4.0.

aaronkirkham commented:
Thanks for the feedback guys, I've removed the magic in the Factory now so self-hosted instances of GitLab must be initialized using the method @YahnisElsts sent. I've also updated instructions in the README for GitLab Integration.

YahnisElsts commented on Puc/v4p2/Vcs/GitLabApi.php:
```diff
@@ -0,0 +1,274 @@
+		/**
+		 * @var string Either a fully qualified repository URL, or just "user/repo-name".
+		 */
+		protected $repositoryUrl;
```
This property is already declared in the parent class and this subclass doesn't read or write it. As far as I can tell, there's no need to re-declare it here; please remove it.

YahnisElsts commented on Puc/v4p2/Vcs/GitLabApi.php:
```diff
@@ -0,0 +1,274 @@
+<?php
+
+if ( ! class_exists( 'Puc_v4p2_Vcs_GitLabApi', false ) ) :
```
Code style: This project doesn't follow the WordPress coding standards on space usage. To stay consistent with the rest of PUC code, please don't put spaces around the list of function arguments and unary operators. This line should look like this:
```php
if ( !class_exists('Puc_v4p2_Vcs_GitLabApi', false) ):
```
There are a bunch of other function calls with too many spaces.

YahnisElsts commented on Puc/v4p2/Vcs/GitLabApi.php:
```diff
@@ -0,0 +1,274 @@
+		 */
+		protected $accessToken;
+
+		public function __construct( $repositoryUrl, $accessToken = null ) {
```
No spaces around the parameter list in function declarations, please.

aaronkirkham commented:
Changes made. All spaces which do not follow the project code style are now removed.
